### ISSUE TITLE
Native crashes do not include codeBundleId. 

### ISSUE THREAD

**Issue by cpetzel** [bugsnag/bugsnag-react-native#267]
### Description

Crashes that occur in native code do not include certain data points from the `configuration` object in javascript. 

### Issue

We use codepush to deliver bundles to our native app. We initialize the bugnsag client with a default `configuration` object, and then update it later when the codepush metatadata comes in 

```CodePush.getUpdateMetadata() ==> which returns JS bundle id```

Then we set the property on the config object. 
```bugsnag.config.codeBundleId```

This works great for errors thrown in Javascript. We see the `codeBundleId` for that error that is thrown. 

However, for native crashes, this `codeBundleId` is missing from the error. This is because the JS Bugsnag layer never tells Native about the new property and therefore it never gets reported during a native crash. (usually the default error handler re-throws the error as a native exception, and it gets reported again)

We could try and reinitialize the bugsnag client via `new Client(config)`, but I have some concerns about calling that constructor more than once...

https://github.com/bugsnag/bugsnag-react-native/blob/master/src/Bugsnag.js#L43-L59

If this method is called more than once, you will be unnecessarily setting the global `ErrorUtils errorHandler` to another instance of a bugsnag handler. 

- What happens to the original error handler? 
- Wouldn't you receive duplicate error reports? (the last registered error handler holds a reference to the previous handler (which is also bugsnag), who in turn holds a reference to the original error handler (which is the original default from React Native). 

**Is there an easier way to tell native about configuration changes? I don't think calling the native init methods is a great solution.** 

### Environment

Library versions:

```shell
├─ bugsnag-react-native@2.10.1
├─ react-native-code-push@5.3.5
└─ react-native@0.55.4

```

- cocoapods version (if any) (`pod -v`): 1.5.3
- iOS/Android version(s): both
- simulator/emulator or physical device?: physical
- debug mode or production?: production

- [x] (iOS only) `[BugsnagReactNative start]` is present in the
  `application:didFinishLaunchingWithOptions:` method in your `AppDelegate`
  class?
- [x] (Android only) `BugsnagReactNative.start(this)` is present in the
  `onCreate` method of your `MainApplication` class?

### Example code snippet

Here is our file that handles all things related to bugsnag. 

```js


import { Client, Configuration } from "bugsnag-react-native";
import { createSelector } from "reselect";
import { getUserEmail, getUserFullName, getUserId } from "../../selectors/user";

import { cloneAndRemoveEmpty } from "../method-helpers";

let bugsnag = null;

function identifyUser() {
  identifyUserWithDetails(reduxStore.getState());
}

export function initBugsnag() {
  if (bugsnag) {
    console.log("bugsnag is already initialized. will return");
    return;
  }

  const configuration = new Configuration();
  configuration.autoCaptureSessions = true;
  configuration.consoleBreadcrumbsEnabled = true;

  configuration.registerBeforeSendCallback((report, error) => {
    if (__DEV__) {
      console.log(
        "Bugsnag caught an error, but we are a debug build, so we will not send the error to bugsnag",
        error
      );
      return false;
    }

    const sanitizedStore = sanitizeStore();

    report.metadata = {
      ...report.metadata, // merge existing metadata that could have been added by other callbacks
      sanitizedStore
    };
    return true;
  });
  bugsnag = new Client(configuration);

  console.log(
    "[Bugsnag] - initBugsnag(). Will subscribe to the root.store"
  );
  reduxStore.subscribe(identifyUser);
  identifyUser();
}

export function setCodePushRelease(codeBundleId) {
  if (bugsnag) {
    console.log("setting the bugsnag codepush config == ", codeBundleId);
    bugsnag.config.codeBundleId = codeBundleId;
  }
}

export function notifyHandledError(error) {
  if (!bugsnag) {
    console.error(
      "[Bugsnag] - trying to notify an error before bugsnag is initialized",
      error
    );
    return;
  }
  if (typeof error === "string") {
    bugsnag.notify(new Error(error));
  } else {
    bugsnag.notify(error);
  }
}

/**
 * Removes properties from the redux store that are too big, or bugsnag reports won't care about
 */
function sanitizeStore() {
  const sanitizedStore = cloneAndRemoveEmpty(reduxStore.getState());
  delete sanitizedStore.locations;
  return sanitizedStore;
}

let loggedIn = false;
const identifyUserWithDetails = createSelector(
  [getUserEmail, getUserFullName, getUserId],
  (email, name, id) => {
    if (email && name && id && bugsnag) {
      console.log("Bugsnag - identifying the user", email, name, id);
      bugsnag.setUser(`${id}`, name, email);
      loggedIn = true;
    } else if (loggedIn && bugsnag) {
      try {
        bugsnag.clearUser();
      } catch (err) {
        console.error("Error trying to log out of Bugsnag");
      }
      loggedIn = false;
    }
  }
);
```
</details>

**Comment by fractalwrench** [comment]
Thanks for the detailed report @cpetzel. I believe the best way forward would be to expose a new API on the JavaScript side for setting the codeBundleId at a later time, similar to this:

```
setCodeBundleId = (id) => {
    NativeClient.setCodeBundleId(id) // add the codeBundleId to Android/iOS metadata
}
```

Would this address your use case?

**Comment by cpetzel** [comment]
Thanks for the response. That would be perfect!

**Comment by nabati** [comment]
What's the status on this, nine months later? We need such a feature as well. 

We went around it by extending the bugsnag Client and adding a `setCodeBundleId` that reinitialises the client. On iOS this works as expected, but on Android it seems like two clients end up running instead (?) and each error gets reported twice, once with the old codeBundleId and once with the new codeBundleId 😢

**Comment by johnkiely1** [comment]
This has been fixed in ([v7.3.4](https://github.com/bugsnag/bugsnag-js/releases/tag/v7.3.4)) of the Bugsnag react native library. `codeBundleId` [can be set](https://docs.bugsnag.com/platforms/react-native/react-native/configuration-options/#codebundleid) in the JS, and once set subsequent native events will include it.

As of v7.3 the library itself has now become part of the [bugsnag-js](https://github.com/bugsnag/bugsnag-js) repository.
An upgrade guide can be found [here](https://github.com/bugsnag/bugsnag-js/blob/next/packages/react-native/UPGRADING.md) with full documentation available [here](https://docs.bugsnag.com/platforms/react-native/react-native/).
